src/rulesets/mergers/exceptions.ts

```diff
@@ -0,0 +1,88 @@
+import { extractPointerFromRef, extractSourceFromRef, pointerToPath } from '@stoplight/json';
+import { isAbsolute, join, normalize as pathNormalize } from '@stoplight/path';
+import { RulesetExceptionCollection } from '../../types/ruleset';
+
+export class InvalidUriError extends Error {
+  constructor(message: string) {
+    super(message);
+  }
+}
+
+const normalize = ($ref: string, rulesetUri?: string): string => {
+  const source = extractSourceFromRef($ref);
+
+  if (typeof source !== 'string') {
+    throw new InvalidUriError(buildInvalidUriErrorMessage($ref, rulesetUri, 'Missing source'));
+  }
+
+  if (rulesetUri === void 0 && source !== '<STDIN>' && !isAbsolute(source)) {
```

Comment:
@P0lip What would be a good file location to define and export that const `<STDIN>` marker and avoid magic strings littering the code?

Reply:
Good question.
I'd say `src/document.ts`

Reply:
@P0lip Fixed